h2o-algos/src/main/java/hex/glm/TweedieVariancePowerMLEstimator.java

```diff
@@ -0,0 +1,1777 @@
+        _totalRows += mrt._totalRows;
+    }
+
+    void cleanSums() {
+        _logVSum = 0;
+        _logVDpSum = 0;
+        _logVDpDpSum = 0;
+        _logVMax = 0;
+        _logWMax = 0;
+        _wSum = 0;
+        _wDpSum = 0;
+        _wDpDpSum = 0;
+    }
+
+    void calculateWjSums(double y, double w) {
+        assert y > 0;
+        final double negwy = (-1) * pow(w, _invp1) * pow(y, 2 * _invp1);
+        final double denom_W_dp_dp_exped = pow(_p1, 2 * _invp1) * _p2 * pow(_phi, _invp1) * pow(y, _p * _invp1);
+        final double log_p1_phi_wy = log(_p1 * _phi / (w * y));
+        final double logs_sumWdp = -(2 - log(_p1 * _phi / (w * y))) * _p + 3 + 2 * log(w * y / (_p1 * _phi));
+        final double p1alphaw = pow(_p1, _alpha) * pow(w, _invp1);
+        final double ps_sumWdp = _p1sq * _p2;
+        final double pphiy_sumWdp = -_p2 * pow(_phi, _invp1) * pow(y, _alpha);
+        final double log_y = log(y);
+        final double log_w = log(w);
+
+        boolean WjLB = false;
+        boolean WjUB = false;
+        boolean WjDpLB = !_needDp;
+        boolean WjDpUB = !_needDp;
+        boolean WjDpDpLB = !_needDpDp;
+        boolean WjDpDpUB = !_needDpDp;
+
+        cleanSums();
+
+        final long jMax = max(2, (long) Math.ceil(w * pow(y, 2 - _p) / ((2 - _p) * _phi)));
+        long j;
+        int cnt = 0;
+        // Start at the maximum and spread out until the change is negligible
+        while (!(WjLB && WjUB && WjDpLB && WjDpUB && WjDpDpLB && WjDpDpUB) && cnt < _max_iter_cnt) {
+            j = jMax + cnt;
+            if (!WjUB) WjUB = Wj(log_y, log_w, j);
+            if (!WjDpUB) WjDpUB = WjDp(logs_sumWdp, p1alphaw, ps_sumWdp, pphiy_sumWdp, j);
+            if (!WjDpDpUB) WjDpDpUB = WjDpDp(y, w, negwy, denom_W_dp_dp_exped, log_p1_phi_wy, j);
+
+            j = jMax - cnt - 1;
+            if (j < 1) {
+                WjLB = true;
+                WjDpLB = true;
+                WjDpDpLB = true;
+            } else {
+                if (!WjLB) WjLB = Wj(log_y, log_w, j);
+                if (!WjDpLB) WjDpLB = WjDp(logs_sumWdp, p1alphaw, ps_sumWdp, pphiy_sumWdp, j);
+                if (!WjDpDpLB) WjDpDpLB = WjDpDp(y, w, negwy, denom_W_dp_dp_exped, log_p1_phi_wy, j);
+            }
+            cnt++;
+        }
+    }
+
+    private boolean Wj(double log_y, double log_w, long j) {
+        final double expInner = ((1 - _alpha) * j) * log_w +
+                (_alpha * j) * (_logp1 - log_y) -
+                j * (1 - _alpha) * _log_phi - j * _logp2 - logGamma(j + 1) - logGamma(-j * _alpha);
+        if (_logWMax == 0) {
+            _logWMax = expInner;
+        }
+        double wSumInc;
+        wSumInc = exp(expInner - _logWMax);
+
+        _wSum += wSumInc;
+        if (_needDp || _needDpDp)  // can use more precision in grad and hess
+            return (abs(wSumInc) + _epsilon) / (abs(_wSum) + _epsilon) < _epsilon && expInner - _logWMax < -37;
+        return expInner - _logWMax < -37;
```

Comment:
The "magic constant" was chosen to have results comparable to R.
It was suggested in [1] and it is used in R's implementation:
> Here $\epsilon = 10^{-16}$ would ensure double precision accuracy in 64-bit floating point arithmetic. In practice we use $\epsilon =\exp(37) \approx 8 \times 10^{17}$

[1]DUNN, Peter K. and SMYTH, Gordon K., 2005. Series evaluation of Tweedie exponential dispersion model densities. Statistics and Computing.